core/src/main/java/org/mskcc/cbio/portal/servlet/PathologyReportView.java

```diff
@@ -0,0 +1,224 @@
+        } catch (NoSuchFileException e) {
+            logger.error("Got No Such File Exception while processing request for pathology report.", e);
+            response.sendError(HttpServletResponse.SC_NOT_FOUND, 
+                "The requested pathology report could not be located.");
+        }
+    }
+    
+    private String getRequestedPath(HttpServletRequest request) throws IOException{
+        return URLDecoder.decode(request.getPathInfo().substring(1), "UTF-8");
+    }
+    
+    private File getRequestedFile(HttpServletRequest request) throws IOException {
+        // Convert from the expected url path 'studyId/patientId/sampleId.pdf' to the expected filesystem path 'studyId.patientId.sampleId.pdf'
+        final String convertedPath = getRequestedPath(request).replace("/", ".");
+        return new File(DATA_DIRECTORY, convertedPath);
+    }
+    
+    private void setError(HttpServletRequest request, int httpStatusCode, String errorMessage) {
+        request.setAttribute(ERROR_CODE, Integer.toString(httpStatusCode));
+        request.setAttribute(ERROR_MSG, errorMessage);
+    }
+    
+    private boolean validateRequest(HttpServletRequest request) throws IOException, DaoException {
+        final String requestedPath = getRequestedPath(request);
+        final String[] path = requestedPath.split("/");
+        if (path.length != 3) {
+            setError(request, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Unable to process requested pathology " +
+                "report path. Ensure it is in the format cbioportal/pathology_report/study_id/patient_id/sample_id.pdf");
+            return false;
+        }
+
+        final String cancerStudyId = path[0];
+        CancerStudy cancerStudy = DaoCancerStudy.getCancerStudyByStableId(cancerStudyId);
+        if (cancerStudy == null) {
+            setError(request, HttpServletResponse.SC_NOT_FOUND, "No such cancer study with id '" + cancerStudyId + "'.");
+            return false;
+        }
+        if (accessControl.isAccessibleCancerStudy(cancerStudyId).size() != 1) {
+            setError(request, HttpServletResponse.SC_FORBIDDEN, 
+                "You are not authorized to access the cancer study with id '" + cancerStudyId + "'.");
+            return false;
+        }
+        
+        final String patientId = path[1];
+        Patient patient = DaoPatient.getPatientByCancerStudyAndPatientId(cancerStudy.getInternalId(), patientId);
+        if (patient == null) {
+            setError(request, HttpServletResponse.SC_NOT_FOUND, "No such patient with id '" + patientId + "' " +
+                "within the cancer study with id '" + cancerStudyId + "'");
+            return false;
+        }
+        
+        final String sampleId = FilenameUtils.removeExtension(path[2]);
+        Sample sample = DaoSample.getSampleByCancerStudyAndSampleId(cancerStudy.getInternalId(), sampleId, false);
+        if (sample == null) {
```

Comment:
They are used more for a sanity check than anything else. When the request comes in it has the URL `www.example.com/cbioportal/pathology_report/study_id/patient_id/sample_id.pdf`. The code uses the study id to check if the user is authorised to the study, and then uses the study_id, patient_id and sample_id when coming up with the file name to get `report_root/study_id.patient_id.sample_id.pdf`. These checks validate the ids provided in the url correspond to actual items in the db, to ensure only the pathology report files are returned (rather than any file within the directory). 

The root directory is also checked for null to ensure that we are only going to be retrieving files from within the pathology report root directory rather than the filesystem root directory.